move to a more standard DSL for F# HTML

dsyme opened the pull request:
This adjusts the HTML DSL for F# as described in https://github.com/dotnet/interactive/issues/687

Also includes #688 to trim back on default namespaces

Typical F# html code looks like this

```fsharp
open Html  // opens Microsoft.DotNet.Interactive.FSharp.FSharpKernelHelpers.Html

let register (df:DataFrame) (writer:TextWriter)
    let take = 20
    table [] [
      thead [] [
        th [] [i [ str "index"]]
        for c in df.Columns -> th [] [c.Name]
      ]
      tbody [] [
        for i in 0 .. min take (int df.Rows.Count - 1) do
          td [] [ object i ]
          for o in df.Rows.[int64 i] do
            td [] [ object o ]
      ]
    ]
    |> writer.Write
Formatter<DataFrame>.Register( (fun df writer -> register df writer), mimeType = "text/html")
```

TODO:
* [x] please check design
* [x] check the notebooks I adjusted by hand

dsyme commented:
This is ready

jonsequitur commented on src/Microsoft.DotNet.Interactive.FSharp.Tests/ApiTests.fs:
```diff
@@ -3,35 +3,39 @@
+
+    [<Fact>]
+    member __.``inner HTML from object``() =
+        Assert.Equal("<div>11</div>", (div [] [object 11]).ToString())
```
A more interesting example here would be an object of a type that leads to additional formatting.

dsyme replied:
I've added quite a number of more tests here

dsyme commented:
I went through all the F# docs for formatting to make sure things were adding up.

Along the way I merged the "HTML.ipynb" into the "Displaying Outputs" and also renamed the notebook files so they don't have spaces in their names, to help me get the cross-references right.

I also added a fix and testing for #703.

dsyme commented:
I also relented on a little on the question of which namespaces are open for F# and also included these by default:

    open System
    open System.IO
    open System.Text

I'll add this to the documentation.

dsyme commented:
OK, I've added a document on the F# notebook programming model.

dsyme commented:
@jonsequitur @colombod ready again

colombod commented on src/Microsoft.DotNet.Interactive.FSharp/FSharpHtml.fs:
```diff
@@ -0,0 +1,511 @@
+    [<AutoOpen>]
+    module HtmlElements =
+        type internal IMarker = interface end
+        let private f name (attrs: HtmlAttribute list) (content: IHtmlContent list) = 
```
could we call ```f``` something like ```htmlElement``` or anything that could make it easier for contributors to identify?

dsyme replied:
Will do.  I will also remove the tags "html" and "body" which we never use in this context to reduce namespace pollution

dsyme replied:
Thanks, done

jonsequitur commented on src/Microsoft.DotNet.Interactive.FSharp/FSharpKernelExtensions.fs:
```diff
@@ -17,35 +17,53 @@ open XPlot.Plotly
+        kernel
+
+    [<Extension>]
+    static member UseExtraNamespacesForTesting(kernel: FSharpKernelBase) =
```
It looks like this should go in the test project.

dsyme replied:
Thanks, done

dsyme commented:
I've adjusted for code review (would love to get this one in as I'm needing it in documentation etc.)
